### PR TITLE
Add Windows ARM64 port to libpgmath

### PR DESCRIPTION
**Pull request by tdusnoki**

### CODE REVIEW

File: runtime/libpgmath/tools/tmp-mth_statsdefs_ARM64.sh [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+			sub("^MTH_DISPATCH_FUNC\\(", "", f); \
+			sub("\\).*", "", f); next; \
+		} \
+		/^[\t ]*_MTH_I_STATS_INC/ { \
+			split($0, s, "[(,)]"); \
+			print "DO_MTH_DISPATCH_FUNC(" f ", " s[2] \
+				", " s[3] ", ", s[4] ")"; f=""; \
+		}' \
+	$1/mth_128defs.c \
+	> $2
```
Comment (xoviat) [inline runtime/libpgmath/tools/tmp-mth_statsdefs_ARM64.sh:25]: There's no difference between this file and the aarch64 file, and they both represent the same architecture. The windows ARM64 should use the same prefix as the linux file.

File: runtime/libpgmath/lib/CMakeLists.txt [hunk 2]
```diff
@@ -6,7 +6,7 @@
 
 # This value will be the same as LIBPGMATH_SYSTEM_PROCESSOR
 set(PROCESSOR ${LIBPGMATH_SYSTEM_PROCESSOR})
-if(${LIBPGMATH_WITH_GENERIC} OR (NOT ${LIBPGMATH_SYSTEM_PROCESSOR} MATCHES "x86_64|aarch64"))
+if(${LIBPGMATH_WITH_GENERIC} OR (NOT ${LIBPGMATH_SYSTEM_PROCESSOR} MATCHES "x86_64|aarch64|ARM64"))
```
Comment (xoviat) [inline runtime/libpgmath/lib/CMakeLists.txt:9]: Come to think of it, why is  the ARM64 introduced at all? The CMake script should check for `aarch64 AND WIN32` or similar.

**Review by xoviat — APPROVED** [review]
utACK

File: runtime/libpgmath/lib/common/acos/fma3/sdacos.cpp [hunk 3]
```diff
@@ -8,7 +8,7 @@
 
 #if defined(TARGET_LINUX_POWER)
 #include "xmm2altivec.h"
-#elif defined(TARGET_LINUX_ARM64)
+#elif defined(TARGET_ARM64)
```
Comment (bryanpkc) [inline runtime/libpgmath/lib/common/acos/fma3/sdacos.cpp:11]: Is `TARGET_ARM64` defined when building on Linux?

File: runtime/libpgmath/lib/common/amod.c [hunk 4]
```diff
@@ -7,11 +7,11 @@
 
 #include "mthdecls.h"
 
-#ifndef WIN64
+#ifndef TARGET_WIN_X8664
 float
 __mth_i_amod(float f, float g)
 {
-#if defined(WIN64)
+#if defined(TARGET_WIN_X8664)
```
Comment (bryanpkc) [inline runtime/libpgmath/lib/common/amod.c:14]: This block is never compiled, due to the outer `#ifndef TARGET_WIN_X8664`. Could you just delete it?

Comment (xoviat) [reply runtime/libpgmath/lib/common/amod.c:14]: looks like it was disabled in 6bc8b5aa55b16afdf0d53fcca024b8aa3bfb4814. Interesting.

File: runtime/libpgmath/lib/common/acos/fma3/sdacos.cpp (hunk 3 above)
Comment (tdusnoki) [reply runtime/libpgmath/lib/common/acos/fma3/sdacos.cpp:11]: It is currently not, good catch thanks. It should be defined IMHO. Any thoughts on that?

Comment (bryanpkc) [reply runtime/libpgmath/lib/common/acos/fma3/sdacos.cpp:11]: Thanks.

File: runtime/libpgmath/lib/common/dmod.c [hunk 5]
```diff
@@ -7,12 +7,12 @@
...
 #include "mthdecls.h"
 
-#ifndef WIN64
+#ifndef TARGET_WIN_X8664
 double
 __mth_i_dmod(double f, double g)
 {
 /* Need to do this way until a bug in the Win64 fmod routine is fixed */
-#if defined(WIN64)
+#if defined(TARGET_WIN_X8664)
```
Comment (bryanpkc) [inline runtime/libpgmath/lib/common/dmod.c:15]: Similar to the amod.c case, this block is also never built.

**Comment by kaadam** [comment]
@bryanpkc @kiranchandramohan What do you think it is fine to be landed?

File: runtime/libpgmath/tools/mth_mask.awk [hunk 6]
```diff
@@ -160,6 +160,29 @@ function init_target_arrays()
...
+    sqrtsd["fs"] = _mm "_sqrt_ps((" __m ")x)"
+    sqrtsd["fd"] = _mm "_sqrt_pd((" __m "d)x)"
+    sqrtsd["rs"] = _mm "_sqrt_ps((" __m ")x)"
+    sqrtsd["rd"] = _mm "_sqrt_pd((" __m "d)x)"
+    sqrtsd["ps"] = _mm "_sqrt_ps((" __m ")x)"
+    sqrtsd["pd"] = _mm "_sqrt_pd((" __m "d)x)"
+    mask_all_zero = \
+    "(_mm_movemask_epi32(_mm_cmpeq_epi32((__m128i)mask, " \
+    "_mm_xor_si128((__m128i)mask,(__m128i)mask))) == 15)"
+
```
Comment (bryanpkc) [inline runtime/libpgmath/tools/mth_mask.awk:185]: This is adding new code to the Linux AArch64 build. While libpgmath seems to build fine with this addition, it is not clear what this is doing, and whether we have enough test coverage to be sure that this change doesn't break something on Linux. Can you change the awk script to run this block only for the Windows build?

File: runtime/libpgmath/lib/common/CMakeLists.txt [hunk 7]
```diff
@@ -157,11 +176,9 @@ else() # Generic Architecture
     mth_128defs_init.c
     mth_128defs_stats.c)
   libmath_add_object_library("${MTH_INTRINSICS_SRCS}" "${FLAGS}" "${DEFINITIONS}" "mth_intrinsics")
- 
+
   set(SRCS
-    kidnnt.c
-    sincos.c
-    sincosf.c)
+    kidnnt.c)
```
Comment (bryanpkc) [inline runtime/libpgmath/lib/common/CMakeLists.txt:181]: Why is this change necessary?

Comment (tdusnoki) [reply runtime/libpgmath/lib/common/CMakeLists.txt:181]: It was for debugging purposes only, thanks for pointing it out, reverting those deletions asap.

**Review by bryanpkc — APPROVED** [review]
Looks OK to me now. Thanks.

**Review by pawosm-arm — APPROVED** [review]

**Comment by kaadam** [comment]
@bryanpkc @kiranchandramohan  What do you think it is okay to be merged?

**Comment by kiranchandramohan** [comment]
I will have a look (end of day for me) and merge if there are no issues.
For windows patches we do not need a signoff from AMD, we only need the CI to pass, but CCing @shivaramaarao anyway.

**Comment by kaadam** [comment]
@kiranchandramohan okay, thanks Kiran.

**Comment by kiranchandramohan** [comment]
Have a few conflicts in our compiler. Need some more time.

File: runtime/libpgmath/CMakeLists.txt [hunk 8]
```diff
@@ -50,6 +50,8 @@ set(LIBPGMATH_SYSTEM_PROCESSOR "${CMAKE_SYSTEM_PROCESSOR}")
 if ("${LIBPGMATH_SYSTEM_PROCESSOR}" STREQUAL "AMD64" OR
     "${LIBPGMATH_SYSTEM_PROCESSOR}" STREQUAL "amd64" )
     set(LIBPGMATH_SYSTEM_PROCESSOR "x86_64")
+elseif ("${LIBPGMATH_SYSTEM_PROCESSOR}" STREQUAL "ARM64")
```
Comment (kiranchandramohan) [inline runtime/libpgmath/CMakeLists.txt:53]: Why was this not required before?

File: runtime/libpgmath/lib/aarch64/cos/gdcos2.c [hunk 9]
```diff
@@ -5,6 +5,7 @@
  *
  */
 
+#if defined(TARGET_LINUX_ARM64)
```
Comment (kiranchandramohan) [inline runtime/libpgmath/lib/aarch64/cos/gdcos2.c:8]: Where will the code below come from for windows?

File: runtime/libpgmath/test/lit.cfg [hunk 10]
```diff
@@ -78,6 +78,9 @@ if config.architecture == "x86_64":
     elif config.operating_system == 'Linux':
         config.test_cflags += " -DTARGET_LINUX_X8664"
 
+if config.operating_system == 'Windows':
+    config.test_cflags += " -DTARGET_WIN"
+
```
Comment (kiranchandramohan) [inline runtime/libpgmath/test/lit.cfg:83]: There is a check for windows above (line 70) where the function append_dynamic_library_path is called. Can this flag addition be moved there?

File: runtime/libpgmath/test/CMakeLists.txt [hunk 11]
```diff
@@ -61,7 +61,7 @@ if(${LIBPGMATH_STANDALONE_BUILD})
   endif()
   set(LIBPGMATH_LIT_ARGS "${LIBPGMATH_LIT_ARGS_DEFAULT}" CACHE STRING
     "Default options for lit")
-  separate_arguments(LIBPGMATH_LIT_ARGS)
+  separate_arguments(LIBPGMATH_LIT_ARGS NATIVE_COMMAND)
```
Comment (kiranchandramohan) [inline runtime/libpgmath/test/CMakeLists.txt:64]: This needs CMake 3.9. Can you increase the minimum requirement where it is appropriate?

File: runtime/libpgmath/lib/common/fltfenv.c [hunk 12]
```diff
@@ -507,7 +507,7 @@ __fenv_feraiseexcept(int exc)
 {
   return feraiseexcept(exc);
 }
-
+#if !defined(TARGET_WIN_ARM64)
```
Comment (kiranchandramohan) [inline runtime/libpgmath/lib/common/fltfenv.c:510]: Where will these functions come from in Windows?

File: runtime/libpgmath/lib/common/dmod.c [hunk 13]
```diff
@@ -7,15 +7,10 @@
 
 #include "mthdecls.h"
 
-#ifndef WIN64
+#ifndef TARGET_WIN_X8664
 double
 __mth_i_dmod(double f, double g)
 {
-/* Need to do this way until a bug in the Win64 fmod routine is fixed */
```
Comment (kiranchandramohan) [inline runtime/libpgmath/lib/common/dmod.c:14]: Was this deadcode?

**Review by kiranchandramohan — CHANGES_REQUESTED** [review]
I have a few questions and comments. I am running a few internal tests as well and will report if I see any issues.

File: runtime/libpgmath/CMakeLists.txt (hunk 8 above)
Comment (xoviat) [reply runtime/libpgmath/CMakeLists.txt:53]: This is required because CMAKE sets the system processor to ARM64 on windows and aarch64 on linux. It's the same processor, so it makes sense to harmonize them here.

File: runtime/libpgmath/lib/CMakeLists.txt [hunk 14]
```diff
@@ -146,15 +146,23 @@ elseif(${LIBPGMATH_SYSTEM_PROCESSOR} MATCHES "ppc64le")
...
-  set(FLAGS_L1 "${FLAGS_L1} -ffp-contract=fast")
+  if(${LIBPGMATH_SYSTEM_NAME} MATCHES "Windows")
+    set(DEFINITIONS_L1 HOST_WIN MAXCPUS=256 MAXCPUSL=8 MAXCPUSR=8 TARGET_WIN
+    TARGET_WIN_ARM64 WIN_ARM64 WINNT WIN64 TARGET_ARM64 _PGI_NOBUILTINS)
+
+    set(DEFINITIONS_L2 HOST_WIN MAXCPUS=256 MAXCPUSL=8 MAXCPUSR=8 TARGET_WIN
+      TARGET_WIN_ARM64 WIN_ARM64 WINNT WIN64 TARGET_ARM64 _PGI_NOBUILTINS)
+  else()
+    set(DEFINITIONS_L1 HOST_LINUX MAXCPUS=8192 MAXCPUSL=128 MAXCPUSR=128
+      TARGET_LINUX TARGET_LINUX_ARM TARGET_LINUX_ARM64 TARGET_ARM64 TARGET_LLVM
```
[the remark this answers is not in the document]
Comment (tdusnoki) [reply runtime/libpgmath/lib/CMakeLists.txt:157]: There are cases when we need to add the same headers on linux and windows. To make it simpler I merged cases when we needed both TARGET_LINUX_ARM64 and TARGET_WIN_ARM64.

File: runtime/libpgmath/lib/common/dmod.c (hunk 13 above)
Comment (tdusnoki) [reply runtime/libpgmath/lib/common/dmod.c:14]: As @bryanpkc pointed it out earlier this was never compiled due to the outer `#ifndef WIN64` .

File: runtime/libpgmath/lib/CMakeLists.txt (hunk 14 above)
[the remark this answers is not in the document]
Comment (kiranchandramohan) [reply runtime/libpgmath/lib/CMakeLists.txt:157]: Here the code is not unified, we have separate cases for windows and others. It is possibly better to remove the TARGET_ARM64 here.

[the remark this answers is not in the document]
Comment (tdusnoki) [reply runtime/libpgmath/lib/CMakeLists.txt:157]: Should I completely separate TARGET_LINUX_ARM64 and TARGET_WIN_ARM64 then? Because if it's not defined here then where should I define it instead?

[the remark this answers is not in the document]
Comment (kiranchandramohan) [reply runtime/libpgmath/lib/CMakeLists.txt:157]: ok. I understand now.

**Comment by kiranchandramohan** [comment]
@tdusnoki will you be updating the patch and replying to the remaining comments?

File: runtime/libpgmath/lib/common/fltfenv.c (hunk 12 above)
Comment (tdusnoki) [reply runtime/libpgmath/lib/common/fltfenv.c:510]: > @tdusnoki will you be updating the patch and replying to the remaining comments?

Yes its currently WIP, I'll update it asap.

**Comment by tdusnoki** [comment]
@kiranchandramohan Yes its currently WIP, I'll update this PR asap.

File: runtime/libpgmath/lib/common/fltfenv.c (hunk 12 above)
Comment (tdusnoki) [reply runtime/libpgmath/lib/common/fltfenv.c]: This needs to be implemented on Windows as these functions are not included in fenv.h. Currently I'm testing possible workarounds and implementations for it.

**Comment by kaadam** [comment]
@kiranchandramohan Tibor updated the change. He left some TODO comment where Windows an ARM specific implementation will be necessary. What do you think it would be good to merge it, and do these implementation in a follow up change?

**Comment by kiranchandramohan** [comment]
@kaadam I am fine with windows specific implementation being done later. Will have a look sometime this week and run some tests and if OK will merge.

**Review by kiranchandramohan — APPROVED** [review]
LGTM.

**Comment by kiranchandramohan** [comment]
Merging as per policy for Windows PRs.
